### ISSUE TITLE
Bit function variant of Universal Approximation Theorem

### ISSUE THREAD

**Issue by quitehappy** [Felix-Petersen/difflogic#18]
Hi,

**Edit: fix operation names (XOR -> EQ)**

Thanks for your great work on DiffLogic. Not sure if this is the right place to discuss, but I think DiffLogic is underrated.

In "C.3 Ternary and Other Additional Operators", you discuss about generalizing 2-to-1 bit functions into n-to-1 bit functions, I believe it leads to a bit function variant of the Universal Approximation Theorem through DiffLogic.

Consider paraphrasing and generalizing the difflogic operator with the following insights:

- Any arbitrary n-to-1 bit function is essentially a Look-Up Table (LUT) with n input address bits and 2^n constant scalar values.
- The combinational logic circuit of a LUT essentially implements a **selective weighted sum** over potential candidate values in the table, where only the 'selected address' carries a nonzero weight.
- This 'selected address' is effectively activated by a **product** of the outputs from a bitwise EQ operation comparing an LUT address with the input address bit by bit.

For clarity, denote the following notation:

- **Vectors** are denoted by uppercase letters, e.g., **`X`** represents a vector.
- **k-th input** within the dataset, which are vectors themselves, are indicated by a uppercase letter with a subscript, e.g., **`X_k`** refers to the input side of the k-th training data point.
- **Scalar Entries** within a vector are denoted by lowercase letters with a subscript, e.g., `x_i` signifies the i-th entry of vector **`X`**.

Given any ground truth bit function `g`:
- Assume the input address bit vector as **X**, or `[x_1, x_2, ..., x_n]`.
- Assume the constant vector representing the LUT data entries as **C**, or `[c_0, c_1, c_2, ..., c_{2^n-1}]`. Alternatively, each scalar value `c_i` can also be denoted in binary indices like `c_{0...00}, ..., c{1...11}`.

Then, using our insight from the LUT, the function `g` can be formulated as:

(1)
<img width="580" alt="g definition" src="https://github.com/Felix-Petersen/difflogic/assets/161915208/0fc6973b-ef98-4640-815d-ec7ede8673fb">

Here, `EQSwitch` is defined by:

<img width="319" alt="EQSwitch" src="https://github.com/Felix-Petersen/difflogic/assets/161915208/cd17bee5-9f8d-4f91-97da-7a6e93a6e81e">

while `b_{ij}` indicates the `j-th` bit in the binary enumeration of index `i`, clarifying whether `x_j` is to be negated or preserved as per the `EQSwitch` logic.

Following the concepts introduced in your paper, we can construct a differential, learnable function **f** by relaxing **C** to a parameterized real-valued vector **W** within `[0, 1]^(2^n)`:

(2)
<img width="653" alt="f definition" src="https://github.com/Felix-Petersen/difflogic/assets/161915208/4091aa9a-1734-44c6-8757-f77e6143d46b">

What's interesting is that when considering MSE loss, we can demonstrate that the loss function with respect to the adjustable parameters **W** is **convex**.

We can start by defining the loss function for a single data point on input **`X_k`** as:
`loss(X_k; W) = (f(X_k; W) - g(X_k; C))^2`.

- Since the input address `X_k` is fixed and the LUT represents a constant table, `g` must be a constant.
- From formula (2), we have `f` is a weighted sum of `w_i`; in other words, `f` is a linear map from the vector **`W`** to a real value.

Let `h(;W) = f(;W) - g`. Given that `g` is a constant, it follows that `h` is also a linear map from the vector **`W`** to a real value.

Then we can prove `loss(;W) = h(;W)^2` is convex, because:
- Any square function is convex, and
- [the composition of a convex function with a linear map is also convex](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Convex/Function.html#ConvexOn.comp_linearMap)

When we consider `Loss(W)` for the entire training dataset as the sum of `loss(X_k; W)` over every single data point, we can deduce that `Loss(W)` is also convex, because [the sum of convex functions is also convex](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Convex/Function.html#ConvexOn.add).

This means that any arbitrary n-to-1 bit function can be learned by a unified difflogic framework when considering MSE loss.

Given the nature of n-to-m bit functions, which can be constructed by concatenating the outputs of m independent n-to-1 bit functions, it also means that any arbitrary n-to-m bit function is learnable by the DiffLogic framework.

In other words, DiffLogic is probably very underrated. **Every widely used network architecture implemented by matrix multiplication today might be able to be reimplemented in DiffLogic**, which potentially leads to **remarkable performance acceleration** during inference time, as you demonstrated in the MNIST experiment.

~~I presume the conclusion holds true when considering other loss functions like cross-entropy; however, I'm unsure of the difficulty in proving it.~~
Edit: I believe the convexity of applying cross-entropy loss on the generalized DiffLogic operator can be proved in a similar way, details omitted.

I hope you find the argument interesting, and I look forward to further developments in DiffLogic.
